internal/controller/apiextensions/composite/api.go

```diff
@@ -265,6 +289,46 @@ func (r *APILabelSelectorResolver) SelectComposition(ctx context.Context, cp res
 	return errors.Wrap(r.client.Update(ctx, cp), errUpdateComposite)
 }
 
+func (r *APILabelSelectorResolver) resolveCustomCompositionSelectors(cp resource.Composite) error {
+	if r.customCompositionSelector == nil {
+		return nil
+	}
+	existingSelector := cp.GetCompositionSelector()
+	if existingSelector != nil && r.customCompositionSelector.GetResolvePolicy() != xpv1.ResolvePolicyAlways {
+		return nil
+	}
+	customSelector := &metav1.LabelSelector{
+		MatchLabels: map[string]string{},
+	}
+
+	for i, m := range r.customCompositionSelector.MatchLabels {
+		key, value, err := r.resolveCustomCompositionSelectorLabel(cp, m)
+		if err != nil {
+			return errors.Wrapf(err, errResolveCustomCompositionSelectorLabel, i)
+		}
+		if key != "" {
+			customSelector.MatchLabels[key] = value
+		}
+	}
+	cp.SetCompositionSelector(customSelector)
+	return nil
+}
+
+func (r *APILabelSelectorResolver) resolveCustomCompositionSelectorLabel(cp resource.Composite, label v1.CustomCompositionSelectorLabel) (string, string, error) {
+	cpMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(cp)
+	if err != nil {
+		return "", "", err
+	}
+	value, err := fieldpath.Pave(cpMap).GetString(label.Value.FromFieldPath)
+	if err != nil {
+		if fieldpath.IsNotFound(err) && label.Value.GetFromFieldPathPolicy() == v1.FromFieldPathPolicyOptional {
+			return "", "", nil
+		}
+		return "", "", err
+	}
+	return label.Key, value, nil
```

Comment:
nit: code style/return early.

```suggestion
	if fieldpath.IsNotFound(err) && label.Value.GetFromFieldPathPolicy() == v1.FromFieldPathPolicyOptional {
		return "", "", nil
	}
	if err != nil {
		return "", "", errors.Wrap(err, errGetRequiredValueFromFieldPath)
	}
	return label.Key, value, nil
```